apps/vscode/src/providers/editor/editor.ts

```diff
@@ -139,8 +143,52 @@ export class VisualEditorProvider implements CustomTextEditorProvider {
       }
     }));
 
+    context.subscriptions.push(window.tabGroups.onDidChangeTabs(async (t) => {
+      const tabs = t.opened;
+
+      if (tabs.length > 0) {
+        for (const tab of tabs) {
+          if (tab.label.endsWith(".qmd") && (tab.input instanceof TabInputText || tab.input instanceof TabInputCustom)) {
+            // determine what mode editor should be in
+            const uri = tab.input.uri;
+
+            const isTextEditor = tab.input instanceof TabInputText;
+            const viewType = isTextEditor ? "textEditor" : tab.input.viewType;
+
+            // get file contents
+            const fileData = await workspace.fs.readFile(uri);
+            const fileContent = Buffer.from(fileData).toString('utf8');
+            const editorMode = determineMode(fileContent, uri);
+            let isSwitch = this.visualEditorPendingSwitchToSource.has(uri.toString()) || this.editorPendingSwitchToVisual.has(uri.toString());
+            if (this.editorPendingSwitchToVisual.has(uri.toString())) {
+              this.editorPendingSwitchToVisual.delete(uri.toString());
+            }
+
+            if (editorMode && editorMode != viewType && !isSwitch) {
+              const allTabs = window.tabGroups.all?.[0]?.tabs;
+
+              // find tab to close if swapping editor type. we don't want to close an active
+              // tab since a tab we are opening has not been set as active yet. we also don't
+              // want to close preview tabs since they will automatically be overriden
+              const tabsToClose = allTabs.filter(tab =>
```

Comment:
I believe we are expecting exactly 1 tab to match here, so you could use `find()` instead of `filter()` and only do the close/open dance when we `find()` a tab, like

```ts
const tabToClose = allTabs.find(stuffhere);

if (tabToClose) {
  await window.tabGroups.close(tabToClose, true);
  await commands.executeCommand("vscode.openWith", uri, editorMode);
}
```

Reply:
That might also be _slightly_ safer in some weird edge cases. Like with the "double open" bugs I've reported in this review, we would not have actually opened the visual editor if we programmed it this way, because `tabToClose` would have been `undefined` and the entire branch would not have run (it still would have been a bug on our part, just a slightly less bad one).